keps/sig-cli/kubectl-socket-support.md

```diff
@@ -0,0 +1,195 @@
+persisted on disk. It is not safe and controllable, especially when this kubeconfig file with credentials gets spread to everywhere.
+Moreover, it will become difficult to revoke the access to the apiserver.
+
+Sockets are used as a secured proxy into some infrastructure. For now kubectl does not expose a way to send the request
+over a unix socket. Currently we've only got `http` and `https`. `Docker` client does provide such supports via three
+different types of Socket: `unix`, `tcp`, and `fd`.
+
+It would be great to be able to send requests from kubectl over sockets.
+
+### Goals
+
+* kubectl users can use sockets to connect to the apiserver:
+    * Add an experimental flag `--experimental-socket-addr` for kubectl, much like `-H unix:///var/run/docker.sock -H tcp://192.168.59.106` in `docker` client;
+    * Allow specifying sockets as server addresses in kubeconfig file;
+
+### Non-Goals
+
+* Convey credentials through unix sockets;
+* Adopting [SGX] to secure the memory access;
+
+[SGX]: https://www.intel.com/content/www/us/en/architecture-and-technology/software-guard-extensions.html
+
+## Proposal
+
+### User Stories
+
+Currently `kubectl` supports running proxy on a unix socket.
+
+```bash
+export TOKEN=USER1_TOKEN
+kubectl proxy --unix-socket=/home/user1/.kube/kube.sock --token=${TOKEN}
+```
+
+Issuing commands to any kube apiserver whereby unix socket is the preferred way to access it.
+
+```bash
+kubectl --experimental-socket-addr=/home/user1/.kube/kube.sock get pods
+```
+
+### Implementation Details/Notes/Constraints
+
+A `--experimental-socket-addr` flag will be added to `kubectl`, much like `-H unix:///var/run/docker.sock -H tcp://192.168.59.106` in `docker` client.
+
+```go
+// staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go
+type ConfigFlags struct {
+    ...
+    
+    // config flags
+    ...
+    Username         *string
+    Password         *string
+    Timeout          *string
+    SocketAddr       *string // +experimental
+    
+    ...
+}
+
+// AddFlags binds client configuration flags to a given flagset
+func (f *ConfigFlags) AddFlags(flags *pflag.FlagSet) {
+    ...
+    if f.SocketAddr != nil {
+        flags.StringVar(f.SocketAddr, "--experimental-socket-addr", *f.SocketAddr, "Path to the socket path/address to use for CLI requests.")
+    }
+}
+```
+
+Also socket addresses can be specified as `Cluster.Server`.
+
+```go
+// staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go
+// Cluster contains information about how to communicate with a kubernetes cluster
+type Cluster struct {
+    // Server is the address of the kubernetes cluster (https://hostname:port).
+    // Socket address is still supported in experimental, such as unix:///var/run/kube.sock, tcp://192.168.59.106, fd://3
+    Server string `json:"server"`
+    ...
+}
+```
+
+To connect to apiserver via sockets, we also need to modify current `TransportConfig`.
```

Comment:
Describe the interaction between credential methods (TLS and bearer token) and socket-based communication. Would TLS credentials be allowed in combination with socket-based communication? Does setting a custom dialer break config caching or certificate rotation?

Reply:
The motivation describes a desire to avoid putting credentials in the kubeconfig file... does that mean they would be disallowed or ignored for socket-based communication? If so, how would the apiserver authenticate the request being made via the socket? (what would the username/uid/groups be set to?)

Reply:
I think the intended use is, kubectl A runs a proxy under A's credentials; kubectl B accesses the proxy. All requests show as A's credentials. This works today (not sure why you do would do this, but it works) as long as you are willing to serve HTTP on localhost.

Reply:
> kubectl A runs a proxy under A's credentials; kubectl B accesses the proxy. All requests show as A's credentials.

Correct. Currently this KEP will only focus on allowing kubectl using unix sockets to connect to the apiserver.

Just as described in non-goals part, the interaction of credentials (tokens, tls) and unix sockets will be not considered.

Quoted from @soltysh ,
> I was expecting this to be only about kubectl support for sockets, the proposal in the current form touches apiserver so kubernetes/sig-api-machinery-feature-requests should be involved too.